Inline Help/Help Center: use the same useHelpSearchQuery hook

yoavf opened the pull request:
The default search results for the inline help are not currently loading: 

<img width="513" alt="Screenshot 2023-06-08 at 13 18 01" src="https://github.com/Automattic/wp-calypso/assets/844866/49b13a07-0b5d-4679-bab1-5161b2376b9f">

Reported in p9Jlb4-7c4-p2#comment-8216


## Proposed Changes

* This updates calypso useHelpSearchQuery hook to use the new endpoint (see https://github.com/Automattic/wp-calypso/pull/76734/ )
* It then removes the custom hook in the Help Center, and uses the calypso hook instead.
* Because the `/help` screen search results need data that the new endpoint does not provide, this also brings back the old hook as `-v1` and uses it for this screen.

## Testing Instructions

* In calypso home, make sure you see the default search results in the help widget:

<img width="500" alt="Screenshot 2023-06-08 at 13 19 32" src="https://github.com/Automattic/wp-calypso/assets/844866/dc59f226-59dc-4a07-ab10-c72205ffd49d">

* In `/help` search results should include both Documentation and Forum links

* The Help Center default results (and search results) should stay the same.

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

yoavf committed:
update calypso's useHelpSearchQuery to use the new endpoint

yoavf committed:
in helpcenter, use calypso's useHelpSearchQuery (which is the same as we had here

yoavf commented:
@alshakero : Does this make sense to you? I didn't want to just duplicate the hook in Calypso. 
I'll test this on atomic/JP of course

alshakero commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,74 @@
+		onSuccess: async ( data ) => {
+			if ( ! data[ 0 ]?.content ) {
+				const newData = await Promise.all(
+					data.map( async ( result: SearchResult ) => {
+						const article: { [ content: string ]: string } = canAccessWpcomApis()
+							? await wpcomRequest( {
+									path: `help/article/${ result.blog_id }/${ result.post_id }`,
+									apiNamespace: 'wpcom/v2/',
+									apiVersion: '2',
+							  } )
```
Pretty rare case, but I think the `async/await` is complicating things here:

```js
Promise.all(data.map(result =>
	wpcomRequest({
		path: `help/article/${result.blog_id}/${result.post_id}`,
		apiNamespace: 'wpcom/v2/',
		apiVersion: '2',
	}).then(article => ({ ...result, content: article.datum }))
))
```

alshakero commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,74 @@
+		},
+		refetchOnWindowFocus: false,
+		refetchOnMount: true,
+		enabled: true,
```
Enabled by default
```suggestion
```

flootr replied:
the same is true for `refetchOnMount`

alshakero reviewed (changes requested):
Thanks for the ping! I think this goes the opposite direction. We should [default to packages, unless we can't](p4TIVU-arC-p2). 

I'd improve the Help Center hook and use it in Calypso. Packages are typed and accessible from both places without pulling any extraneous dependencies.

When you import a package, you import only that, when you import something from Calypso, you really don't know what you're pulling with it.

yoavf committed:
reverse the hook location

Instead of the hook being in calypso and imported in help-center, we do the
opposite.

yoavf committed:
progress?

gravityrail committed:
Clean up API code

yoavf committed:
Change how inline-search-results gets its data following the API change

flootr commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,90 @@
+	return useQuery< any >( {
+		queryKey: [ 'help-center-search', search, sectionName ],
+		queryFn: fetchArticlesAPI,
+		onSuccess: async ( data ) => {
```
Please take into consideration that `onSuccess` callbacks for queries (not mutations!) are [deprecated](https://tanstack.com/query/v5/docs/react/guides/migrating-to-v5#callbacks-on-usequery-and-queryobserver-have-been-removed) and will be removed in the next major version of `@tanstack/query` so it may be worthwhile to explore different solutions to fetching additional content.

simison commented:
If the PR works, the the feedback here can probably addressed in a follow-up PR? The important thing would be getting a bug fixed sooner and optimization can wait.

simison reviewed (approved):
Looks to work well:

<img width="720" alt="image" src="https://github.com/Automattic/wp-calypso/assets/87168/4b733bf2-3182-4e5f-9209-98b3c7f04378">

yoavf committed:
Merge remote-tracking branch 'origin/trunk' into update/useHelpSearchQuery

yoavf committed:
bring back the old version of the hook, since the new version isn't compatible with the
'/help' screen search results

yoavf commented:
@simison 
> Looks to work well:

Yes, but then the `/help` screen was broken because it expects a different format for the data:
<img width="1099" alt="Screenshot 2023-06-18 at 12 33 33" src="https://github.com/Automattic/wp-calypso/assets/844866/e5c9699a-a44b-4a02-8ed3-ccb087019b47">

To solve this asap I brought back the old hook in `use-help-search-query-v1.ts`.

alshakero commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,90 @@
+		locale
+	) }&section=${ encodeURIComponent( sectionName ) }`;
+	if ( canAccessWpcomApis() ) {
+		fetchArticlesAPI = () =>
```
I would make `fetchArticlesAPI` a function that has these `ifs` and `elses` inside it. In which case, you can put it outside the hook to avoid recreating it with every render.

```js
function fetchArticlesAPI() {
	if ( canAccessWpcomApis() ) {
		return wpcomRequest( {
				path: `help/search/wpcom${ params }`,
				apiNamespace: 'wpcom/v2/',
				apiVersion: '2',
		} );	
	} else {		
		return apiFetch( {
				global: true,
				path: `/help-center/search${ params }`,
			} as APIFetchOptions );
	}
}
```

yoavf replied:
Thanks!  implemented

alshakero commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,90 @@
-			wpcomRequest( { path: '/help/search', query: params.toString(), apiVersion: '1.1' } ),
-		enabled: !! search,
+	// set up article fetching
+	let fetchArticleAPI: ( blogId: number, postId: number ) => Promise< { content: string } >;
+	if ( canAccessWpcomApis() ) {
+		fetchArticleAPI = ( blogId: number, postId: number ) =>
+			wpcomRequest( {
+				path: `help/article/${ blogId }/${ postId }`,
+				apiNamespace: 'wpcom/v2/',
+				apiVersion: '2',
+			} );
+	} else {
+		fetchArticleAPI = ( blogId: number, postId: number ) =>
+			apiFetch( {
+				global: true,
+				path: `/help-center/fetch-post?blog_id=${ blogId }&post_id=${ postId }`,
+			} as APIFetchOptions );
+	}
```
+1.

alshakero commented on client/data/help/use-help-search-query.ts:
```diff
@@ -1,26 +1,90 @@
+				} )
+			);
+
+			queryClient.setQueryData( [ 'help', search ], newData );
```
Does this query key exist? I can't find it. If you're trying to alter the results of the query itself (and not another query), [select](https://tanstack.com/query/v4/docs/react/reference/useQuery#:~:text=select%3A%20(data%3A%20TData)%20%3D%3E%20unknown) is better, IMO.

yoavf replied:
@alshakero Not sure I understand (this isn't something I've changed here - it's already in [the existing hook](https://github.com/Automattic/wp-calypso/blob/trunk/packages/help-center/src/hooks/use-help-search-query.ts))

alshakero replied:
What this line does, it sets the data of a query that has the same key. Eg

```js
useQuery( [ 'help', search ], ... );
```

And I searched Calypso and it doesn't exist, so I imagine it's a leftover. It's probably not doing anything here.

yoavf replied:
Thanks @alshakero  - cleaned that up!

alshakero reviewed (changes requested):
Found one issue with the code. Will test in a following iteration.

yoavf committed:
extract callbacks for fetching articles and a single article

alshakero reviewed (approved):
Works and looks great! Thank you so much for picking this up.

As for the stale query key, it's up to you.

yoavf committed:
Remove leftover queryClient code
